File: litfass-server/src/main/scala/lit/fass/server/script/GroovyScriptEngine.scala
package lit.fass.server.script

import java.net.{URL, URLClassLoader}
import java.time.ZoneOffset.UTC
import java.time.format.DateTimeFormatter.ISO_DATE_TIME
import java.util

import groovy.lang.GroovyClassLoader
import javax.script.ScriptEngineManager
import lit.fass.server.ScriptLanguage
import lit.fass.server.ScriptLanguage.ScriptLanguage
import lit.fass.server.model.{Data, DataList}
import lit.fass.server.script.GroovyScriptEngine._
import org.apache.commons.lang3.time.StopWatch
import org.slf4j.LoggerFactory

object GroovyScriptEngine {
  private val lang = ScriptLanguage.GROOVY
  private val log = LoggerFactory.getLogger(classOf[GroovyScriptEngine])
  private val scriptLog = LoggerFactory.getLogger(s"$lang.Script")
  private val scriptTimestampFormatter = ISO_DATE_TIME.withZone(UTC)
  private val classPath: Array[String] = System.getProperty("java.class.path")
    .split(":")
    .toList
    .filter(it => !it.contains("*") && (it.contains("groovy") || it.contains("slf4j-api") || it.contains("ext/")))
    .toArray

  def apply(): GroovyScriptEngine = {
    log.debug(s"Using classpath $classPath")
    new GroovyScriptEngine()
  }
}

/**
 * @author Michael Mair
 */
class GroovyScriptEngine extends ScriptEngine {

  override def isApplicable(language: ScriptLanguage): Boolean = lang == language

  override def invoke(script: String, data: DataList): DataList = {
    log.trace(s"Invoking script:\n$script\nwith data \n$data")
    val result = DataList()

    val stopWatch = StopWatch.createStarted()

    var classLoader: GroovyClassLoader = null
    try {
      classLoader = new GroovyClassLoader(new URLClassLoader(classPath.map(it => new URL(s"file:$it"))))
      val scriptEngineManager = new ScriptEngineManager(classLoader).getEngineByName(lang.toString.toLowerCase)
      val bindings = scriptEngineManager.createBindings()
      bindings.put("log", scriptLog)
      bindings.put("timestampFormatter", scriptTimestampFormatter)
      bindings.put("data", if (data.size == 1) data.get(0) else data)
      val evalResult = scriptEngineManager.eval(script, bindings)

      evalResult match {
        case _: util.Collection[util.Map[String, Object]] =>
          evalResult.asInstanceOf[util.Collection[util.Map[String, Object]]].forEach(it => result.add(Data.of(it)))
        case _: util.Map[String, Object] =>
          result.add(Data.of(evalResult.asInstanceOf[util.Map[String, Object]]))
        case _ =>
          log.error(s"Unable to process evaluation result of type ${evalResult.getClass.getName}")
      }
    } finally {
      classLoader.close()
    }

    stopWatch.stop()
    log.debug(s"Script executed in ${stopWatch.formatTime()}")
    result
  }
}

File: litfass-server/src/main/kotlin/lit/fass/server/script/groovy/GroovyScriptEngine.kt
package lit.fass.server.script.groovy

import  groovy.lang.GroovyClassLoader
import lit.fass.server.script.ScriptEngine
import lit.fass.server.script.ScriptLanguage
import lit.fass.server.script.ScriptLanguage.GROOVY
import org.apache.commons.lang3.time.DurationFormatUtils.formatDurationHMS
import org.slf4j.LoggerFactory
import java.net.URL
import java.net.URLClassLoader
import java.time.ZoneOffset.UTC
import java.time.format.DateTimeFormatter.ISO_DATE_TIME
import javax.script.ScriptEngineManager
import kotlin.system.measureTimeMillis

/**
 * @author Michael Mair
 */
class GroovyScriptEngine : ScriptEngine {
    companion object {
        private val lang = GROOVY
        private val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
        private val scriptLog = LoggerFactory.getLogger("$lang.Script")
        private val scriptTimestampFormatter = ISO_DATE_TIME.withZone(UTC)
    }

    private val classPath: List<String> = System.getProperty("java.class.path")
        .splitToSequence(":")
        .filter {
            !it.contains("*") && (
                    it.contains("groovy") ||
                            it.contains("slf4j-api") ||
                            it.contains("ext/")
                    )
        }
        .toList()

    init {
        log.debug("Adding classpath $classPath")
    }

    override fun isApplicable(language: ScriptLanguage): Boolean {
        return lang == language
    }

    override fun invoke(script: String, data: Collection<Map<String, Any?>>): Collection<Map<String, Any?>> {
        log.trace("Invoking script:\n$script\nwith data \n$data")
        var result: Collection<Map<String, Any?>> = listOf(emptyMap())
        val elapsedTime = measureTimeMillis {
            GroovyClassLoader(URLClassLoader(classPath.map { URL("file:$it") }.toTypedArray())).use { classLoader ->
                result = with(ScriptEngineManager(classLoader).getEngineByName(lang.name.toLowerCase())) {
                    val evalResult = eval(script, createBindings().apply {
                        put("log", scriptLog)
                        put("timestampFormatter", scriptTimestampFormatter)
                        if (data.size == 1) {
                            put("data", data.first())
                        } else {
                            put("data", data)
                        }
                    })

                    when (evalResult) {
                        is Collection<*> -> {
                            @Suppress("UNCHECKED_CAST")
                            evalResult as Collection<Map<String, Any?>>
                        }
                        is Map<*, *> -> {
                            @Suppress("UNCHECKED_CAST")
                            listOf(evalResult as Map<String, Any?>)
                        }
                        else -> emptyList()
                    }
                }
            }
        }
        log.debug("Script executed in ${formatDurationHMS(elapsedTime)}")
        return result
    }
}